BUG: fix categorical comparison with missing values (#26504 )

another-green opened the pull request:
- [x] closes #26504
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

This PR fixes issues comparison of ordered categorical with missing values evaluates to True(#26504 ).  Now making comparison with missing values always gives False. The missing values could be introduced when constructing a Categorical with values that don't fall in with the categories provided.

mitstake committed:
ENH - Index set operation modifications to address issue #23525 (#23538)

simonjayhawkins committed:
DOC/CLN: wil -> will (#26484)

javabrett committed:
Fixed typo mutiplication -> multiplication. (#26489)

simonjayhawkins committed:
DOC: fix SyntaxError in doc build on Windows (#26499)

matsmaiwald committed:
DOC: Highlighted role of index alignment in DataFrame.dot(other) (#26480) (#26496)

simonjayhawkins committed:
DOC/CLN: Change API reference section title (#26486)

topper-123 committed:
CLN: Remove StringMixin from PandasObject (#26505)

vaibhavhrt committed:
Fix type annotations in pandas.core.indexes.datetimes (#26404)

another-green committed:
BUG: None comparison evaluates to True #26504

matsmaiwald committed:
Better error message for DataFrame.hist() without numerical columns (… (#26483)

WillAyd committed:
Excel Test Cleanup - ReadWriteClass (#26473)

pep8speaks commented:
Hello @yanglinlee! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/tests/arrays/categorical/test_operators.py`](https://github.com/pandas-dev/pandas/blob/cf156cff6ea7544b5ad57b70f8927eda240da44a/pandas/tests/arrays/categorical/test_operators.py):

> [Line 199:52](https://github.com/pandas-dev/pandas/blob/cf156cff6ea7544b5ad57b70f8927eda240da44a/pandas/tests/arrays/categorical/test_operators.py#L199): [E241](https://duckduckgo.com/?q=pep8%20E241) multiple spaces after ','

gfyoung commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,6 +186,18 @@ def test_comparison_with_unknown_scalars(self):
+        cat = Categorical([1, 2, 3, None], categories=[1, 2, 3], ordered=True)
+
+        msg = ("Cannot compare a Categorical for op __{}__ with a scalar,"
+               " which is not a category")
```
What this `msg` variable for?

yanglinlee replied:
I was trying to return some error messages and got interrupted by my coworkers. Basically, it should not be there.

gfyoung commented:
@yanglinlee : Thanks for submitting this!  Can you also add a `whatsnew` entry to `v0.25.0.rst?`

mroeschke committed:
CLN: pd.TimeGrouper (#26477)

mroeschke committed:
CLN: Remove ExcelWriter.sheetname (#26464)

xref gh-6581

yanglinlee commented:
> @yanglinlee : Thanks for submitting this! Can you also add a `whatsnew` entry to `v0.25.0.rst?`

Sure! Done~

gfyoung commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -305,7 +305,7 @@ Categorical
 
 - Bug in :func:`DataFrame.at` and :func:`Series.at` that would raise exception if the index was a :class:`CategoricalIndex` (:issue:`20629`)
--
+- Bug in :func:`_cat_compare_op` that would valuate comparison with None to True (:issue:`26504`)
```
Our `whatsnew` entries try to describe the bugs from the perspective of user-facing functionality.  In this case, the bug is affecting comparisons between `Categorical` and scalars.  Can you update the entry to describe what's being affected there instead?

yanglinlee replied:
Thank you for the suggestion! Will do!

mroeschke committed:
CLN: Remove deprecated parse_cols from read_excel (#26522)

xref gh-6581

makbigc committed:
[TEST] Add two more parameters to the test_dti_add_sub_nonzero_mth_offset (#26392)

* Add two more parameters to the test

* Add array into the boy and add parameter freq

WillAyd reviewed (changes requested):
So to be clear this has nothing to do with the `None` value right? Isn't this generally applicable to values that don't fall into a category and if so can you update whatsnew / commentary and tests to reflect that?

nandahkrishna committed:
Remove py.path special handling from io.common (#26458)

topper-123 committed:
CLN: remove StringMixin from code base, except core.computation (#26523)

jreback reviewed (changes requested):
I am puzzled of the purpose of this PR. We don't' allow *any* comparison of categorical except for eq / ne for non-ordered. I would be ok with simply raising the same TypeError message that we already do. But what is the purpose of this expansion?

xcz011 committed:
MAINT: port numpy#13188 for np_datetime simplification (#26516)

* MAINT: port numpy#13188 for np_datetime simplificaion

Bring numpy changes about emulating the behavior of python's divmod to
pandas.

* cpplint fix

* Add reference numpy change into comment

* fix typo

yanglinlee commented:
> I am puzzled of the purpose of this PR. We don't' allow _any_ comparison of categorical except for eq / ne for non-ordered. I would be ok with simply raising the same TypeError message that we already do. But what is the purpose of this expansion?

Just want to get some clarification: 
Scenario 1: Missing values. 
```
In [50]:
cat = pd.Categorical([1, 2, 3, None], categories=[1, 2, 3], ordered=True) 
cat
Out[50]:
[1, 2, 3, NaN]
Categories (3, int64): [1 < 2 < 3]
In [51]: 
cat <=2
Out[51]:
array([ True,  True, False,  True])
```
If we allow missing values in the array, then the result should not be raising errors.
People suggest the result in Out[51] should be `array([ True,  True, False,  False])`

Scenario 2: Values that don't fall into a category as WillAyd suggested.
```
In [54]:
cat2 = pd.Categorical([1, 2, 3, 999], categories=[1, 2, 3], ordered=True) 
cat2
Out[54]:
[1, 2, 3, NaN]
Categories (3, int64): [1 < 2 < 3]
In [55]:
cat2<=2
Out[55]:
array([ True,  True, False,  True])
```
Currently it doesn't raise TypeError. 

If you want to raise TypeError for both scenarios, I can change the behavior of the comparison with scalar. Let me know your suggestions. Thanks.

jreback commented:
if the above scenarios compare with a listlike they raise i believe
a scalar should not be different

yanglinlee commented:
> if the above scenarios compare with a listlike they raise i believe
> a scalar should not be different

I tested comparison with listlike, and it didn't raise TypeError. It only takes comparison with None as False. 
```
cat1 = pd.Categorical([1,2,3,999], categories=[1,2,3], ordered=True)
cat2 = pd.Categorical([2,2,2,2], categories=[1,2,3], ordered=True)
cat1
Out[]:
[1, 2, 3, NaN]
Categories (3, int64): [1 < 2 < 3]
In[]:
cat1 >= cat2
Out[]:
array([False,  True,  True, False])
```
In this case, do we need to change the behavior of comparison to listlike? So in both cases, the comparison to None will raise Error.

jorisvandenbossche commented:
> I am puzzled of the purpose of this PR.

The purpose is quite clear to me: fixing a bug. Which is clearly described in the issue: https://github.com/pandas-dev/pandas/issues/26504

The bug is that when doing a comparison (which is allowed for ordered categoricals), missing values always evaluate to True or False (depending on which operations, because they are encoded as -1 and we use that value in the comparison). 
While comparisons to NaN should always be False.

To show this in another way:

```
In [49]: cat = pd.Categorical([1, 2, 3, None], categories=[1, 2, 3], ordered=True)                         

In [50]: pd.Series(cat)           
Out[50]: 
0      1
1      2
2      3
3    NaN
dtype: category
Categories (3, int64): [1 < 2 < 3]

In [51]: pd.Series(cat) < 2    
Out[51]: 
0     True
1    False
2    False
3     True
dtype: bool

In [52]: pd.Series(cat, dtype=object) < 2                     
Out[52]: 
0     True
1    False
2    False
3    False
dtype: bool
```

The object dtype has the correct behaviour: a missing value gives False in the result.


> So to be clear this has nothing to do with the None value right?

This is not about "None", or also not about values that don't fall within the category, but is is about *missing values* (but both None as non-existent categories when constructing the categorical lead to missing values in the actual categorical)

jorisvandenbossche commented on pandas/core/arrays/categorical.py:
```diff
@@ -100,7 +100,15 @@ def f(self, other):
+                # check for NaN in self
+                na_mask = (self._codes == -1)
+                if na_mask.any():
+                    # In other series, the leads to False, so do that here too
```
I know you copied this comment from above, but it is actually not fully clear to me what it wants to say, so would just leave it out here.

(or say something that comparison to NaN always gives False)

yanglinlee replied:
Sounds good! Great coaching!

jorisvandenbossche commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,6 +186,19 @@ def test_comparison_with_unknown_scalars(self):
         tm.assert_numpy_array_equal(cat != 4,
                                     np.array([True, True, True]))
 
+    def test_comparison_with_known_scalars(self):
```
can you rename this test to something like `test_compatison_with_missing_values` ?

yanglinlee replied:
Will do!

jorisvandenbossche commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,6 +186,19 @@ def test_comparison_with_unknown_scalars(self):
+        # be evaluated as False
+
+        cat1 = Categorical([1, 2, 3, None], categories=[1, 2, 3], ordered=True)
+        cat2 = Categorical([None, 1, 2, 3], categories=[1, 2, 3], ordered=True)
```
I don't think the two cases are needed?

yanglinlee replied:
I agree.

jorisvandenbossche reviewed (commented):
Generally looking good to me, thanks for the PR!

Can you address the comment of @gfyoung about the whatsnew message?

jreback commented:
this needs to raise for non ordered categoricals

that is not addressed in this PR as there are no tests for this

furthermore the behavior for listlikes must also be the same (raise for non ordered)
if u r making work for scalars then it needs to work for listlikes as well (for ordered)

jorisvandenbossche commented:
> this needs to raise for non ordered categoricals

This does already raise, that is not the issue. 

> furthermore the behavior for listlikes must also be the same. if u r making work for scalars then it needs to work for listlikes as well (for ordered)

This does already work for listlikes. Actually this PR basically copied the code from the code branch for list-likes to the code branch of scalars.

WillAyd commented:
> This is not about "None", or also not about values that don't fall within the category, but is is about _missing values_ (but both None as non-existent categories when constructing the categorical lead to missing values in the actual categorical)

Gotcha - so in both in the original issue and here in the whatsnew we keep referring to `None` as being of importance but that is distracting from the actual issue. 

@yanglinlee looks like you might have some other changes to the whatsnew but can you account for that on next edit? Ping back if you are stuck on wording

jorisvandenbossche commented:
> Gotcha - so in both in the original issue and here in the whatsnew we keep referring to None as being of importance but that is distracting from the actual issue.

Yes, it was by using "None" instead of "missing values" that caused some confusion I think. I already changed the title of this PR.

yanglinlee commented:
Thank everyone for the helpful discussions! This is great suggestion! I will update the PR after work tonight.

another-green committed:
fix categorical comparison with missing values #26504

another-green committed:
Merge branch 'master' into master

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -348,7 +348,7 @@ Categorical
 
 - Bug in :func:`DataFrame.at` and :func:`Series.at` that would raise exception if the index was a :class:`CategoricalIndex` (:issue:`20629`)
--
+- Fixed Bug in :func:`_cat_compare_op` that would evaluate comparison of ordered `Categorical` with missing values with scalar to True sometimes (:issue:`26504`)
```
this needs to be user facing

jorisvandenbossche replied:
```suggestion
- Fixed bug in comparison of ordered `Categorical` that contained missing values with a scalar which sometimes incorrectly resulted in True (:issue:`26504`)
```

(we don't want to mention the `_cat_compare_op`, as that is not relevant for the end user

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -100,7 +100,14 @@ def f(self, other):
+                ret = getattr(self._codes, op)(i)
+
+                # check for NaN in self
+                na_mask = (self._codes == -1)
```
just call this mask

jreback commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,10 +185,20 @@ def test_comparison_with_unknown_scalars(self):
         tm.assert_numpy_array_equal(cat != 4,
                                     np.array([True, True, True]))
 
+    def test_comparison_of_ordered_categorical_with_missing_values(self):
```
run thru all of the comparison ops here with the compare_operators_no_eq_ne fixture

yanglinlee replied:
Sounds good!

jreback commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,10 +185,20 @@ def test_comparison_with_unknown_scalars(self):
         tm.assert_numpy_array_equal(cat != 4,
                                     np.array([True, True, True]))
 
+    def test_comparison_of_ordered_categorical_with_missing_values(self):
```
add the word scalar in here

yanglinlee replied:
Done

jreback commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,10 +185,20 @@ def test_comparison_with_unknown_scalars(self):
+        # and following comparisons with scalars in categories with missing
+        # values should be evaluated as False
+
+        cat = Categorical([1, 2, 3, None], categories=[1, 2, 3], ordered=True)
```
also compare this with a list-like (can be separate test or same)

yanglinlee replied:
Added as a separate test case

another-green committed:
Update test_operators.py

another-green committed:
Update categorical.py

another-green committed:
Update test_operators.py

another-green committed:
Update test_operators.py

jorisvandenbossche reviewed (commented):
@yanglinlee Thanks for the updates. Added a few more comments.

another-green committed:
Update doc/source/whatsnew/v0.25.0.rst

Co-Authored-By: Joris Van den Bossche <jorisvandenbossche@gmail.com>

yanglinlee commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,10 +185,30 @@ def test_comparison_with_unknown_scalars(self):
+
+        cat = Categorical([1, 2, 3, None], categories=[1, 2, 3], ordered=True)
+
+        assert getattr(cat, compare_operators_no_eq_ne)(2)[-1] == False
```
This will fail, since now we test compare_operators_no_eq_ne which includes <, <=, >, >=. I need to list all four expected results.

jreback replied:
you can do this

```
with warnings.catch_warnings():
            warnings.simplefilter("ignore", RuntimeWarning)
            expected = getattr(np.array(cat), op)(2)
```

which will result in a numpy array (note that we suppress the warnings because numpy does the correct comparison with NaN but it shows a warning)

yanglinlee replied:
Thanks for suggestion, Jeff! I struggled to make pytest.mark.parametrize work.

another-green committed:
Update test_operators.py

another-green committed:
Update test_operators.py

another-green committed:
Update v0.25.0.rst

another-green committed:
Update test_operators.py

another-green committed:
Update test_operators.py

another-green committed:
Update test_operators.py

another-green committed:
Update test_operators.py

yanglinlee commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -348,7 +348,7 @@ Categorical
 
 - Bug in :func:`DataFrame.at` and :func:`Series.at` that would raise exception if the index was a :class:`CategoricalIndex` (:issue:`20629`)
--
+- Fixed bug in comparison of ordered `Categorical` that contained missing values with a scalar which sometimes incorrectly resulted in True (:issue:`26504`)
```
Thanks! Done

yanglinlee commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,10 +185,30 @@ def test_comparison_with_unknown_scalars(self):
+
+        cat = Categorical([1, 2, 3, None], categories=[1, 2, 3], ordered=True)
+        other = Categorical([2, 2, 2, 2], categories=[1, 2, 3], ordered=True)
+        assert getattr(cat, compare_operators_no_eq_ne)(other)[-1] == False
```
Thanks! Done~

jreback commented on pandas/core/arrays/categorical.py:
```diff
@@ -1425,6 +1433,7 @@ def notna(self):
 
         """
         return ~self.isna()
+
```
can u remove these extra lines

yanglinlee replied:
NP

jreback commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,6 +186,36 @@ def test_comparison_with_unknown_scalars(self):
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", RuntimeWarning)
+            actual = getattr(cat, compare_operators_no_eq_ne)(scalar)
+            expected = getattr(np.array(cat),
```
can u pull the expected and assert out of the context only the actual needs to be there

yanglinlee replied:
Yes, good point!

jreback commented on pandas/tests/arrays/categorical/test_operators.py:
```diff
@@ -186,6 +186,36 @@ def test_comparison_with_unknown_scalars(self):
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", RuntimeWarning)
+            actual = getattr(cat, compare_operators_no_eq_ne)(other)
+            expected = getattr(np.array(cat), compare_operators_no_eq_ne)(2)
```
this should be other yes?

yanglinlee replied:
Done

another-green committed:
Update categorical.py

another-green committed:
Update test_operators.py

WillAyd reviewed (approved):
lgtm outside of @jreback comments

TomAugspurger commented on pandas/core/arrays/categorical.py:
```diff
@@ -89,18 +89,25 @@ def f(self, other):
+
+                # check for NaN in self
+                mask = (self._codes == -1)
+                if mask.any():
```
Why do we need `mask.any()`? Can we not just have `ret[mask] = False`

yanglinlee replied:
Hi Tom, it seems the CI config is buggy. I checked several PR have similar errors. Need help~

TomAugspurger replied:
@yanglinlee can you comment on https://github.com/pandas-dev/pandas/pull/26514#discussion_r289191338?

yanglinlee replied:
I think it is a good idea. Let's make it simpler and quicker.

TomAugspurger reviewed (commented):
Not sure what happened with the last CI run. Try merging master and pushing again.

another-green committed:
Merge remote-tracking branch 'upstream/master' into cat

TomAugspurger commented:
@yanglinlee CI is being fixed in https://github.com/pandas-dev/pandas/pull/26586. Merge master & repush once that's in.

another-green committed:
Merge remote-tracking branch 'upstream/master' into cat

another-green committed:
fix categorical comparison with missing values

another-green committed:
Merge remote-tracking branch 'upstream/master' into cat

jreback commented:
thanks @yanglinlee
